homeassistant/components/modbus/fan.py

```diff
@@ -0,0 +1,142 @@
+"""Support for Modbus fans."""
+from datetime import timedelta
+import logging
+from typing import Any, Dict, Optional
+
+from homeassistant.components.fan import FanEntity
+from homeassistant.const import CONF_NAME, CONF_SCAN_INTERVAL
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import (
+    ConfigType,
+    DiscoveryInfoType,
+    HomeAssistantType,
+)
+
+from . import ModbusHub
+from .const import (
+    CALL_TYPE_COIL,
+    CALL_TYPE_REGISTER_HOLDING,
+    CONF_FANS,
+    CONF_REGISTER,
+    CONF_REGISTER_TYPE,
+    MODBUS_DOMAIN,
+)
+from .switch import ModbusCoilSwitch, ModbusRegisterSwitch
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(
+    hass: HomeAssistantType,
+    config: ConfigType,
+    async_add_entities,
+    discovery_info: Optional[DiscoveryInfoType] = None,
+):
+    """Read configuration and create Modbus fan."""
+    if discovery_info is None:
+        return
+
+    fans = []
+    for fan in discovery_info[CONF_FANS]:
+        hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
+
+        if CALL_TYPE_COIL in fan:
+            fans.append(ModbusCoilFan(hub, fan))
+
+        if CONF_REGISTER in fan:
+            fans.append(ModbusRegisterFan(hub, fan))
+
+    async_add_entities(fans)
+
+
+class ModbusCoilFan(ModbusCoilSwitch, FanEntity):
+    """Representation of a Modbus coil fan."""
+
+    def __init__(self, hub: ModbusHub, config: Dict[str, Any]):
+        """Initialize the coil fan."""
+        super().__init__(hub, config)
+        self._scan_interval = timedelta(seconds=config[CONF_SCAN_INTERVAL])
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        state = await self.async_get_last_state()
+        if not state:
+            return
+        self._is_on = state.state
+
+        async_track_time_interval(
+            self.hass, lambda arg: self._update(), self._scan_interval
+        )
+
+    @property
+    def should_poll(self):
+        """Return True if entity has to be polled for state.
+
+        False if entity pushes its state to HA.
+        """
+
+        # Handle polling directly in this entity
+        return False
+
+    # pylint: disable=arguments-differ
```

Comment:
Can we refactor the entity classes in switch and fan platforms and extract the common part to a platform generic class so we don't need to disable pylint here?

Reply:
The problem lies in the `FanEntity` base class. It inherits from `ToggleEntity`, and breaks the interface by introducing one more parameter in the function call:

Relevant code:
```python
    # pylint: disable=arguments-differ
    def turn_on(self, speed: Optional[str] = None, **kwargs) -> None:
        """Turn on the fan."""
        raise NotImplementedError()
```

This issue is therefore related only to Modbus fan. I would love to avoid disabling this pylint check, but because of the base class I don't know how.

If you have some ideas, I can rework the code.

Reply:
This doesn't affect all fan platforms, which leads me to believe that it's the SwitchEntity mixin in the MRO that causes the pylint error here in this platform. I'm not totally sure about that though. My idea was having a cleaner MRO maybe could avoid it.

Reply:
I will take a look at other implementations. If there's a way how to avoid disabling linter, I will do it.